### PR TITLE
Improve recursion depth checks

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR improves relationship checking for recursive types in a number of ways:

* When checking for potentially recursive instantiations of the same type, we now only consider types with higher type IDs. Because type IDs are monotonically increasing, newer type instantiations have higher type IDs than older type instantiations--and in a recursive type that generates an infinite series of type instantiations in some pattern, we therefore continue to see increasing type IDs. One particular advantage of this change is that manually nested type instantiations of the form `Foo<Foo<Foo<Foo<Foo<T>>>>>` are _not_ counted as recursive instantiations (because we instantiate from innermost to outermost).
* We now consider a type to be "deeply nested" when we have seen three distinct instantiations of the type with increasing type IDs. Previously the limit was five, but without the increasing type ID requirement. This change means we'll do less work to detect and stop _infinitely generated_ types, but will check _manually nested_ types to any depth. Previously we couldn't distinguish between the two.
* We improve our method of computing and checking "broadest equivalent IDs" that was introduced in #42727.

The change to computing equivalent IDs improves type checking performance by 2-5%. Our performance test suites don't contain examples of infinitely generated types, so the depth check changes show no effect there. However, several Definitely Typed packages see dramatic improvement. For example, `redux-immutable`, `react-lazylog`, and `yup` all see at least a 50% reduction in check time.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster
@typescript-bot test this
@typescript-bot user test inline
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
@typescript-bot user test this inline

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster
@typescript-bot user test this inline
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster
@typescript-bot run dt

**Comment by ahejlsberg** [comment]
@typescript-bot perf test faster
@typescript-bot run dt

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -20261,47 +20259,55 @@ namespace ts {
...
                 source = target;
                 target = temp;
             }
             const postFix = intersectionState ? ":" + intersectionState : "";
-            if (isTypeReferenceWithGenericArguments(source) && isTypeReferenceWithGenericArguments(target)) {
-                const typeParameters: Type[] = [];
-                return getTypeReferenceId(source as TypeReference, typeParameters) + "," + getTypeReferenceId(target as TypeReference, typeParameters) + postFix;
-            }
-            return source.id + "," + target.id + postFix;
+            return isTypeReferenceWithGenericArguments(source) && isTypeReferenceWithGenericArguments(target) ?
```
Comment (weswigham) [inline src/compiler/checker.ts:20308]: I've been wondering this for awhile and worked on a change for a bit, but can/should we expand this detailed id generation to generic type-alias'd types as well (and not just type references), so type alias relations can also benefit from the broadest-equivalent-id check?

**Review by weswigham — APPROVED** [review]
I like the structure of this - the `broadestEquivalentId` check is ultimately an approximation; ideally we'd be able to note all relations that are strictly a specialization of an ongoing relation check and know to bail, but the id comparison is a pretty cheap way to check something close to that and hits the very common example of recursion during variance computations. Thus, any changes that make us pick up more equivalent ids in a cheap way is generally welcome. The depth detection changes are a bit more worrysome -  doesn't the id-increasing check mean that repeated usages of the exact same type id no longer count as "deeply nested"? I guess the assumption there is that deep nesting that isn't generative in some way must form a cycle at some point which will be detected. Still - definitely has the potential to convert some "quietly works" code into "loudly breaks with a stack depth error" code.

**Comment by ahejlsberg** [comment]
> doesn't the id-increasing check mean that repeated usages of the exact same type id no longer count as "deeply nested"?

Actually no, we still include identical type IDs in the count. We only skip counting when we see a _lower_ type ID. My comments aren't exactly clear on that, I'll revise.

**Comment by weswigham** [comment]
Oh, I know, but we could have a check sequence like 1,2,1,1,1,1,... and previously we'd have marked it as deep (5x repeated 1), whereas because of the 2, now we won't.

**Comment by ahejlsberg** [comment]
Actually, we will mark that as deep because we always compare against the _previous_ type ID, not the _highest_ type ID.

**Comment by weswigham** [comment]
Oh, so I see. I guess that means 1,2,3,4,3,2,1 won't be marked as deep (the last 3,2,1 won't be counted), but 1,2,3,4,1,2,3 will (4 -> 1 won't count, but 1->2 and 2->3 will), even though it's just a reordering of the same types.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.5

**Comment by mkantor** [comment]
I think this fixes #42070.

**Comment by Jym77** [comment]
Hi, I want to point a small regression that I believe has been introduced by this PR 😕 Alas I haven't been able to isolate a MNWE after 1 day trying to extract code from my project and reaching 200 lines in the non so minimal and still working example 😞 

TLDR: [This project](https://github.com/Siteimprove/alfa) builds in TS4.5.2 but doesn't in TS 4.5.3 (nor in 4.5.5).
Given the error message and the crashing on our controlled-recursive type, I do believe that this PR is the cause of it.

We are building recursive type but controlling the recursion through a `type Depths = [-1, 0, 1, 2]` (ensuring `Depths[D] = D-1`), then the type looks like (over simplified) `type Nested<T, D extends number = 3> = D extends -1 ? T : Nested<T, Depths[D]>`. The `D` parameter ensures that type recursion stops…

Of course, [the actual type](https://github.com/Siteimprove/alfa/blob/main/packages/alfa-act/src/interview.ts#L28-L47) is much more crazy. It models an "interview" which is a sequence of questions (asked to an oracle), where each question expects a typed answer (`T` is a representation of this type (e.g. `"boolean"`, `"string[]"`), and `A` is the actual corresponding type (e.g. `boolean`, `Array<string>`); the `QUESTION` parameter is meta-data mapping the questions' URI to their type). An interview may contains several questions in a row (hence the recursion), and the depth is used to make sure the type works.

The type itself does build in TS 4.5.3, but then the way we use it (through yet some other layers of complication) breaks…
I've tried to adapt the code to get it working in 4.5.5. From first attempts, it seems that giving some hints about the actual value of `D` helps 🤔 It feels like TS is now too eager in setting `D=3` and thus struggles when nesting interviews…

I don't really consider this as a bug. Feels more like a very corner case… Plus haven't been able to get a MWNE to study… But I guess you might want to hear about it 😊 
(and my secret hope is that you'll have a "aha!" moment seing the description and knowing the content of this PR much better than me 😀 )
